### ISSUE TITLE
Expired messages in prefetch queue after synchronous calls.

### ISSUE THREAD

**Issue by conniey** [Azure/azure-sdk-for-java#25259]
## Summary:

It's possible to receive stale messages and the "The lock supplied is invalid." when completing a message.
1. We do not renew prefetched messages:
    * while they are in the buffer
    * nor, if they are in the user's received message batch (`IterableStream<ServiceBusMessage>`).
2. While the current thread is processing events, it's possible for additional events to come in, in our background thread and be queued.

## Question:
Is this expected behaviour? Should we change it or document it?

```java
ServiceBusReceiverClient receiver = buildReceiverClient();

// 100 credits are placed on the link because we want 100 messages.
// However, 50 are returned.
var messages = receiver.receive(100, Duration.ofSeconds(10));

// While these messages are being processed, another 50 messages are received 
// from the service and put into the backing prefetch queue.
for (var message : messages) {
   process(message);
}

// We receive another 100 messages. 50 will be served from the prefetch queue
// and we put 50 credits on the link.
// The first 50 messages we served may already be expired because:
// * We do not renew the lock token of messages in prefetch.
messages = receiver.receive(100, Duration.ofSeconds(10));
```

Related #25204 , #25203 , #25121, #20640

## Background:

We never supported auto-lock-renew in LowLevel Sync Receiver (Iterable based). 

We have a work item to remove the error-prone auto-lock-renew from LowLevel Async Receiver (which algins Java with other SDK's low-level client).

So the auto-lock-renew feature is something we don't want to offer in LowLevel (Sync, Async) Receivers. Hence there is always a possibility of the user receiving lock expired messages (from prefetch-queue).

The multiple `receive()` APIs serve messages from common prefetch-queue, which leads to a scenario in which a later `receive()` receives "expired messages" followed by "un-expired messages" (as shown in the above example).

#### Option_1 - Document the behavior:

The simplest option is to document this ("expired messages" followed by "un-expired messages") as expected behavior. i.e. Locks expiring for messages in the prefetch queue is a tradeoff for people opting into using prefetch.

####  Option_2 - drain-flow performative:

The second option is to take advantage of drain-flow performative. The idea is:
  * The `receive(max-messages, timeout)` API returns `Iterable`.
  * An `iterable` is "active" once the user starts iterating it.
  * Once the `Iterable` reaches the `completion-state` (e.g. max-messages consumed, timeout), meaning it should not return more messages, a cleanup phase runs. 
  * The cleanup-phase sent drain-flow, wait for the drain-ack, clear prefetch-queue then `Iterable` enters `terminal-state`.
  * Unless the active Iterable is in the `terminal-state`, we don't allow the user to start iterating `Iterable` obtained from other `receive(..)` API calls. 

## Concerns with Option_2:

There is still a possibility of internally receiving inflight messages between drain-flow-send and drain-flow-ack; we can drop these messages, say (seq: 11-to-20). But the concern is, when users call `receive()` again, is there is a possibility of receiving messages "out of order,", i.e., they receive (seq: 21-to-30) then (seq: 11-to-20) because the dropped (seq: 11-to-20) showed up once it's server-side lock expired.

So the problem of "expired messages" followed by "un-expired messages" turned into "out of order messages".

We can incorporate "abandon" as part of dropping. But the question is it's a lot for the "cleanup-phase" (which requires sending multiple frames to service, the possibility of network drop, retries, etc..) to have a clean slate for the next `receive()` API call.

**Comment by sohanlal311** [comment]
What if the prefetch count is ZERO and **renewMessageLock** method is manually called with a big time (say 48 hours) on all the received messages before processing them? Actually, we are making sure that the locks get continuously renewed for all messages until long running processing is complete.

Please refer this code snippet for the exact scenario: https://github.com/Azure/azure-sdk-for-java/issues/25121#issuecomment-963908826

**Comment by sohanlal311** [comment]
I have analyzed the internal code from azure library and found the code snippet which is causing the issue.

**package:** com.azure.messaging.servicebus;
**class:** LockRenewalOperation implements AutoCloseable {
**method:**    private Flux<OffsetDateTime> getRenewLockOperation(OffsetDateTime initialLockedUntil, Duration maxLockRenewalDuration) {
**Note:** Please look at the lines marked with **----->** in below code.
```
        return Flux.switchOnNext(emitterProcessor.map(interval -> Mono.delay(interval)
            .thenReturn(Flux.create(s -> s.next(interval)))))
            .takeUntilOther(cancellationSignals)
            .flatMap(delay -> {
                logger.info("token[{}]. now[{}]. Starting lock renewal.", lockToken, OffsetDateTime.now());
                return renewalOperation.apply(lockToken);
            })
            .map(offsetDateTime -> {
----->      final Duration next = Duration.between(OffsetDateTime.now(), offsetDateTime);
                logger.info("token[{}]. nextExpiration[{}]. next: [{}]. isSession[{}]", lockToken, offsetDateTime, next,
                    isSession);

----->      sink.next(MessageUtils.adjustServerTimeout(next));
                return offsetDateTime;
            });
```

Here is the code for **MessageUtils.adjustServerTimeout** method which is the second occurrence in above code which is marked with **----->**:
```
    // Pass little less than client timeout to the server so client doesn't time out before server times out
    public static Duration adjustServerTimeout(Duration clientTimeout) {
        return clientTimeout.minusMillis(1000);
    }
```

**My analysis:**
1. The reference **next** is the difference between the current time and the next expiration time. It will be a few milliseconds. Suppose it is 300 milliseconds.
2. Then you are adjusting the time in **MessageUtils.adjustServerTimeout** method by subtracting 1000 milliseconds from the **next** time.

**Conclusion:**
Hence according to the above logic, the next renewal from the client side will roughly happen just 1.3 seconds before the next expiration. Please note it may fail as well due to network latency or some other IO related issues.

**Fix:**
Could you please allow the client to pass a time which will be used in **MessageUtils.adjustServerTimeout** method rather than just 1 second so that client has control to renew the lock. It will really be beneficial in cases where we are fetching a batch of messages and processing them.

**Comment by conniey** [comment]
@sohanlal311 

> I have analyzed the internal code from azure library and found the code snippet which is causing the issue.

Thanks for the analysis. We increased the timeout because from the original client library (com.microsoft.azure:microsoft-servicebus).

https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/primitives/Util.java#L347-L349

> Fix:
Could you please allow the client to pass a time which will be used in MessageUtils.adjustServerTimeout method rather than just 1 second so that client has control to renew the lock. It will really be beneficial in cases where we are fetching a batch of messages and processing them.

If the user wants control over when they're renewing, should they be using manual lock renewal over auto-lock renewal? I agree that the timespan should be updated. I asked our other language components how they figure out when to renew, and .NET uses: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ReceiverManager.cs#L358, which may be a better approach.

**Comment by sohanlal311** [comment]
@conniey 

The fix you have made in 

https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/primitives/Util.java#L347-L349

I can see that you have replaced 1000 with 200 which means the lock will be renews just 200 milliseconds before the next expiration. I think, it will cause more issues because likelihood of not renewing the lock is increased. Please correct me if I am wrong because it seems your version of library doesn't match with the one I am using.

**Comment by conniey** [comment]
> The fix you have made in

https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/primitives/Util.java#L347-L349

The code I reference is where we _originally_ used as a template for our new client libraries. You'll notice that the package name is entirely different. I was trying to provide context for where this logic came from.

We haven't made any fixes on this yet.

**Comment by ramya-rao-a** [comment]
Update regarding the discussion started by @sohanlal311 

- We need to update the logic used to determine when exactly to renew the lock when the auto lock renewal feature is enabled. We will be using [the logic used in the .NET SDK](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ReceiverManager.cs#L358) which is the same as what we use in our TypeScript/JavaScript SDK as well.
- This is currently planned to be fixed as part of our January update i.e in the second week of January 2022

**Comment by vtrusevich-incomm** [comment]
Is there any workaround, that we can implement before we are waiting till the January release?

**Comment by vtrusevich-incomm** [comment]
@ramya-rao-a, please, let us know if there will be any Snapshot/RC versions with the fix applied. We would like to apply it for our system and test on lower environments as well

**Comment by ki1729** [comment]
Changes are done - should be part of the January release - should be released in the next week

**Comment by ramya-rao-a** [comment]
Thanks @ki1729 

@conniey #26517 has been logged to investigate option 2 i.e the drain feature

@ki1729  Lets use this issue to track option 1 which is adding the documentation to clarify expectations

**Comment by vtrusevich-incomm** [comment]
So, @ki1729 , @ramya-rao-a have I understood it correctly, that the fix implemented by @ki1729 won't be released this week (which is actually ended) as was initially planned?

In our case, we are using 0 prefetch count and we have 1-2 messages out of 100 each day that do fail to prolong the locking period. The fix from @ki1729 is exactly solving the issue for us and we delaying our releases waiting for the fix performed by @ki1729

**Comment by sohanlal311** [comment]
@ramya-rao-a I see a new version 7.5.2 on https://mvnrepository.com/artifact/com.azure/azure-messaging-servicebus. Could you please let me know if it includes this fix?

**Comment by vtrusevich-incomm** [comment]
@sohanlal311 , according to release notes here https://github.com/Azure/azure-sdk-for-java/releases/tag/azure-messaging-servicebus_7.5.2, this fix is included into released version

**Comment by ramya-rao-a** [comment]
Hey @vtrusevich-incomm and @sohanlal311 

Yes, 7.5.2 version of azure-messaging-servicebus has the fix around the lock renewal as being discussed in this issue and was released on 14th of Jan, 2022. My bad, I assumed it was being released this week.

**Comment by ramya-rao-a** [comment]
Closing this issue for the below reasons:
- The issue around lock renewal is now fixed and the fix is available in version 7.5.2 of azure-messaging-servicebus
- https://github.com/Azure/azure-sdk-for-java/pull/26447 added the link to the prefetch doc in the docs for receiveMessages()
- We have https://github.com/Azure/azure-sdk-for-java/issues/26517 logged to track the drain feature which is in the backlog 
- On a more immediate front, we have
    -  https://github.com/Azure/azure-sdk-for-java/issues/26632 to investigate if we can release messages without increasing the delivery count if their lock has expired before they were delivered to the user
    - https://github.com/Azure/azure-sdk-for-java/issues/26637 to improve the way we add credits to reduce messages being put into the buffer
-
